src/codegen/compile.ml

```diff
@@ -1164,8 +1168,14 @@ module Stack = struct
   let set_stack_ptr env =
     G.i (GlobalSet (nr (E.get_global env "__stack_pointer")))
 
-  (* TODO: check for overflow *)
   let alloc_words env n =
+    (* first, check for stack underflow *)
+    (* optimize me *)
```

Comment:
Maybe put this behind the `--check` flag.

Reply:
#3090 also suggests a stack guard that never allocates the topmost page.

Reply:
Also only do the check if `n` is so big that it would step over the topmost guard page.

Reply:
> Maybe put this behind the `--check` flag.

If we have some other reliable guard in place yes, but not until then.

Reply:
Sure - all reasonable optimizations, but I think we should always have some form of reliable guard. Let's see if this actually adds to the cost much anyway.